### ISSUE TITLE
False Positive? Mutable key type lint is emitted on HashSets containing associated types

### ISSUE THREAD

**Issue by TimoFreiberg** [rust-lang/rust-clippy#5020]
```
$ cargo clippy -V                                                                     
clippy 0.0.212 (e8642c7 2020-01-06)                                                   
```
Minimal example:
```
trait Trait {
    type AssociatedType;
    
    fn trait_fn(&self, set: std::collections::HashSet<Self::AssociatedType>);
}
```
A `mutable key type` lint is emitted on `std::collections::HashSet<Self::AssociatedType>` in the trait definition, but not in impls of that trait (interestingly not even in trait impls where `Trait::AssociatedType` is `AtomicUsize`).
While `Trait::AssociatedType` could be an internally mutable type, I think the correct place to lint this would be in impls that actually set `Trait::AssociatedType` to an internally mutable type.

Reproducible on the playground: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=094f6136a26c9911ee8dae0dcf94f150
